src/ripple/app/tx/impl/Payment.cpp

```diff
@@ -30,6 +30,19 @@ namespace ripple {
 
 // See https://ripple.com/wiki/Transaction_Format#Payment_.280.29
 
+XRPAmount
+Payment::calculateMaxSpend(STTx const& tx)
+{
+    auto const sendMax = tx[~sfSendMax];
+    auto const saDstAmount(tx.getFieldAmount(sfAmount));
+
+    /* If there's no sfSendMax in XRP, and the sfAmount isn't
+    in XRP, then the transaction can not send XRP. */
+    return sendMax ?
+        (sendMax->native() ? sendMax->xrp() : beast::zero) :
+        (saDstAmount.native() ? saDstAmount.xrp() : beast::zero);
+}
+
```

Comment:
I'm not convinced I got the field accessing interfaces to `STObject` quite right, so if usage "in the field" guides us to possible improvements in how its implemented, I'm on board.